commit 592c7d76246554df8b8771ae0af920f6dc361af7
Author: Peter Rajnoha <prajnoha@redhat.com>
Date:   Fri Mar 18 13:42:06 2016 +0100

    fsadm: if available, use /proc/self/mountinfo to detect mounted volume
---
 WHATS_NEW        |  1 +
 scripts/fsadm.sh | 36 ++++++++++++++++++++++++++++++++----
 2 files changed, 33 insertions(+), 4 deletions(-)

diff --git a/WHATS_NEW b/WHATS_NEW
index cdc6894..a413f45 100644
--- a/WHATS_NEW
+++ b/WHATS_NEW
@@ -1,5 +1,6 @@
 Version 2.02.147 - 
 =====================================
+  If available, use /proc/self/mountinfo to detect mounted volume in fsadm.
   Fix resize of stacked raid thin data volume (2.02.141).
 
 Version 2.02.146 - 
diff --git a/scripts/fsadm.sh b/scripts/fsadm.sh
index fc85988..26be102 100755
--- a/scripts/fsadm.sh
+++ b/scripts/fsadm.sh
@@ -76,6 +76,7 @@ MOUNTPOINT=
 MOUNTED=
 REMOUNT=
 PROCMOUNTS="/proc/mounts"
+PROCSELFMOUNTINFO="/proc/self/mountinfo"
 NULL="$DM_DEV_DIR/null"
 
 IFS_OLD=$IFS
@@ -188,6 +189,14 @@ detect_fs() {
           # hardcoded /dev  since udev does not create these entries elsewhere
 	  /dev/dm-[0-9]*)
 		read </sys/block/${RVOLUME#/dev/}/dm/name SYSVOLUME 2>&1 && VOLUME="$DM_DEV_DIR/mapper/$SYSVOLUME"
+		read </sys/block/${RVOLUME#/dev/}/dev MAJORMINOR 2>&1 || error "Cannot get major:minor for \"$VOLUME\""
+		;;
+	  *)
+		STAT=$(stat --format "MAJOR=%t MINOR=%T" ${RVOLUME}) || error "Cannot get major:minor for \"$VOLUME\""
+		eval $STAT
+		MAJOR=$((0x${MAJOR}))
+		MINOR=$((0x${MINOR}))
+		MAJORMINOR=${MAJOR}:${MINOR}
 		;;
 	esac
 	# use null device as cache file to be sure about the result
@@ -198,11 +207,18 @@ detect_fs() {
 	verbose "\"$FSTYPE\" filesystem found on \"$VOLUME\""
 }
 
-# check if the given device is already mounted and where
-# FIXME: resolve swap usage and device stacking
-detect_mounted()  {
-	test -e "$PROCMOUNTS" || error "Cannot detect mounted device \"$VOLUME\""
+detect_mounted_with_proc_self_mountinfo()  {
+	MOUNTED=$("$GREP" "^[0-9]* [0-9]* $MAJORMINOR " "$PROCSELFMOUNTINFO")
+
+	# extract 5th field which is mount point
+	# echo -e translates \040 to spaces
+	MOUNTED=$(echo ${MOUNTED} | cut -d " " -f 5)
+	MOUNTED=$(echo -n -e ${MOUNTED})
+
+	test -n "$MOUNTED"
+}
 
+detect_mounted_with_proc_mounts()  {
 	MOUNTED=$("$GREP" "^$VOLUME[ \t]" "$PROCMOUNTS")
 
 	# for empty string try again with real volume name
@@ -224,6 +240,18 @@ detect_mounted()  {
 	test -n "$MOUNTED"
 }
 
+# check if the given device is already mounted and where
+# FIXME: resolve swap usage and device stacking
+detect_mounted()  {
+	if test -e "$PROCSELFMOUNTINFO"; then
+		detect_mounted_with_proc_self_mountinfo
+	elif test -e "$PROCMOUNTS"; then
+		detect_mounted_with_proc_mounts
+	else
+		error "Cannot detect mounted device \"$VOLUME\""
+	fi
+}
+
 # get the full size of device in bytes
 detect_device_size() {
 	# check if blockdev supports getsize64
